### PR TITLE
Add callback to catch NaNs in the train loss

### PR DESCRIPTION
**Pull request by coryMosaicML**
Adds a basic NaN catching callback that will raise an error if a NaN in the train loss is found.

To use, add
```
  nancatcher:
    _target_: diffusion.callbacks.NaNCatcher
```
As a callback in your yaml.

### CODE REVIEW

File: diffusion/callbacks/nan_catcher.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from typing import Sequence
+
+import torch
+from composer import Callback, Logger, State
+
+
+class NaNCatcher(Callback):
+    """Catches NaNs in the loss and raises an error if one is found."""
+
+    def after_loss(self, state: State, logger: Logger):
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/callbacks/nan_catcher.py:15]: Have tested SD2, which AFAIK has the same loss output as SDXL. Some models can output a dictionary, but I believe this is handled by composer [here].(https://github.com/coryMosaicML/composer/blob/3830a55fa4676701fe6c4ca6dbbe6ed447e10218/composer/trainer/trainer.py#L2445)

[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/callbacks/nan_catcher.py:15]: Verified that it also works with SDXL.

**Review by Landanjs — APPROVED** [review]
Could you also check if loss is a dictionary? Otherwise, LGTM!

File: diffusion/callbacks/nan_catcher.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        """Check if loss is NaN and raise an error if so."""
+        # Should check if any of the elements of the loss are NaN
+        if isinstance(state.loss, torch.Tensor):
+            if torch.isnan(state.loss).any():
+                raise RuntimeError('Train loss contains a NaN.')
+        elif isinstance(state.loss, Sequence):
+            for loss in state.loss:
+                if torch.isnan(loss).any():
+                    raise RuntimeError('Train loss contains a NaN.')
+        elif isinstance(state.loss, Dict):
```
Comment (Skylion007) [inline diffusion/callbacks/nan_catcher.py:25]: Nit: this could have been an abstract mapping class based on the methods used on it.
